### DISCUSSION TITLE
Setting the correct pts and time_base when using PiCamera2 with aiortc

### DISCUSSION THREAD

**Discussion by hackernese** [aiortc/aiortc#968] [Q&A]
Hi, currently i am using a Raspberry pi with a Module V3 camera ( which forces the usage of picamera2 library ), so i tried to look at your "webcam" example which led me to make some small changes involving adding the following class
```python
from picamera2 import Picamera2
from av import VideoFrame
from fractions import Fraction
from aiortc import MediaStreamTrack

class OverwrittenCamera(MediaStreamTrack):

    kind = "video"

    def __init__(self):
          super().__init__()  # don't forget this!
          self.cam = Picamera2()
          self.cam.configure(self.cam.create_preview_configuration())
          self.cam.start()

    async def recv(self):
          img = self.cam.capture_array()
          # Capture a frame as a numpy array

          # rebuild a VideoFrame, preserving timing information
          new_frame = VideoFrame.from_ndarray(img, format='rgba')
          new_frame.pts = 0                         # NOT SURE WHAT TO PUT HERE
          new_frame.time_base = Fraction(1,1000000) # NOT SURE WHAT TO PUT HERE
          return new_frame
```
and altered the "offer" function a little bit like this 
```python
async def offer(request):
     params = await request.json()
     offer = RTCSessionDescription(sdp=params["sdp"], type=params["type"])

     pc = RTCPeerConnection()
     pcs.add(pc)

     @pc.on("connectionstatechange")
     async def on_connectionstatechange():
          print("Connection state is %s" % pc.connectionState)
          if pc.connectionState == "failed":
               await pc.close()
               pcs.discard(pc)

     # Altered section : 
     # Since i do not need audio but only the video track, i simplified it as follows
     cam = OverwrittenCamera()
     pc.addTrack(cam)

     await pc.setRemoteDescription(offer)

     answer = await pc.createAnswer()
     await pc.setLocalDescription(answer)

     return web.Response(
          content_type="application/json",
          text=json.dumps(
               {"sdp": pc.localDescription.sdp, "type": pc.localDescription.type}
          ),
     )
```
the rest remains the same, but because i can't grab "time_base" and "pts" value ( because "img" is a numpy array ) like some other examples in your repository like so...
```python
class VideoTransformTrack(MediaStreamTrack):

     # kind="video" and __init__(), etc goes here

     async def recv(self):
          frame = await self.track.recv()

          # other if else code....

          # rebuild a VideoFrame, preserving timing information
          new_frame = VideoFrame.from_ndarray(img, format="bgr24")
          new_frame.pts = frame.pts              # <== GRAB the correct pts value here !!!
          new_frame.time_base = frame.time_base  # <== GRAB the correct time_base value here  !!!
          return new_frame
```
Which is why i think that the live streaming process has a significant delay between each frame after I press the 'Start' button. To provide more details, I initially pointed my camera towards a blank wall. Once I exchanged SDP and successfully established a live stream from my Raspberry Pi to the browser, it displayed the blank wall. However, when I moved my camera towards my window, there was a delay of around 3-10 seconds before it showed the new view.

So is there anyhow i can extract the correct information or is there any workaround for this ? Please help.

PS: since i am not sure if this is a problem relating to Raspberry pi specifically or rather aiortc, please redirect me if needed

**Comment by hackernese** [comment]
Ok so after a bit digging and reading, i finally made the live streaming process smooth without any delay by updating the class with the following...
```python
class OverwrittenCamera(MediaStreamTrack):

    kind = "video"

    def __init__(self):
        super().__init__()  # don't forget this!
        self.cam = Picamera2()
        self.cam.configure(self.cam.create_preview_configuration())
        self.timeit = None
        # Create a variable to keep track of the time this track is first added to the stream
        self.cam.start()

    async def recv(self):

        if self.timeit==None:
            # Set the current time
            self.timeit = time.time()

        img = self.cam.capture_array()

        # Calculating the representation time in term of microseconds
        pts = (time.time() - self.timeit) * 1000000

        # rebuild a VideoFrame, preserving timing information
        new_frame = VideoFrame.from_ndarray(img, format='rgba')
        new_frame.pts = int(pts)
        new_frame.time_base = Fraction(1,1000000)
        return new_frame
```
However, i still want to really understand how this works. So in my understanding, the "pts" determines when the frame is allowed to be sent to the user ( browser in this case ) ? and time_base basically represents the time unit of "pts" ? 
But if i try to get rid of `self.timeit` and basically replace `pts = (time.time() - self.timeit) * 1000000` with `pts = time.time() * 1000000`, it still works normally, even though in that case, the `pts` variable will start with 1 microsecond but rather 53 years' worth of microseconds. So, I would appreciate it if anyone could explain why aiortc does not wait for like a quadrillion seconds before sending the frame but instead detects it and sends it right away ?

**Comment by soycaporal** [comment]
I suggest you give this libav guide a read. How PTS and time base are calculated will make more sense.
[https://github.com/leandromoreira/ffmpeg-libav-tutorial#chapter-1---syncing-audio-and-video](https://github.com/leandromoreira/ffmpeg-libav-tutorial#chapter-1---syncing-audio-and-video)
